pkg/pillar/hypervisor/kvm.go

```diff
@@ -462,20 +462,135 @@ func (ctx kvmContext) Task(status *types.DomainStatus) types.Task {
 	}
 }
 
-// TBD: Have a better way to calculate this number.
-// For now it is based on some trial-and-error experiments.
-// Container limit is reduced to 100MiB.
-func minVMMOverhead(config types.DomainConfig) int64 {
+func estimatedVMMOverhead(domainName string, config types.DomainConfig, aa *types.AssignableAdapters) (int64, error) {
+	// Container limit is set to 100MiB.
+	// This is a temporary solution until we have a better way to predict
+	// the memory usage of the container.
 	if config.IsOCIContainer() {
-		return 100 << 20 // Mb in bytes
+		return 100 << 20, nil // Mb in bytes
 	}
-	return 600 << 20 // Mb in bytes
-}
 
-func vmmOverhead(config types.DomainConfig,
-	globalConfig *types.ConfigItemValueMap) (int64, error) {
 	var overhead int64
 
+	mmioOverhead, err := mmioVMMOverhead(domainName, config, aa)
+
+	if err != nil {
+		return 0, logError("mmioVMMOverhead() failed for domain %s: %v",
+			domainName, err)
+	} else {
+		overhead = undefinedVMMOverhead(config) + ramVMMOverhead(config) +
+			qemuVMMOverhead(config) + cpuVMMOverhead(config) + mmioOverhead
+	}
+
+	return overhead, nil
+}
+
+func ramVMMOverhead(config types.DomainConfig) int64 {
+	// 2.5% of the total RAM alocated for VM in bytes
+	// this formula is precise and well explained in the following article:
+	// TBD
+	return int64(config.Memory) * 1024 * 25 / 1000
```

Comment:
@OhmSpectator  I have updated version of this code already. I just realized that 0.25% is wrong! see https://gitlab.com/qemu-project/qemu/-/issues/1003

Reply:
Actually, the estimation of 8 bytes per 4 Kb mentioned there is not right. 8 bytes per PTE is ok, but
1) They are never allocated by one by 8 bytes; they are allocated in bigger chunks.
2) Some extra mem is necessary not only for PTEs but also for the other levels of page tables (PDs, PDPTs).
So, the 0.2-0.25% is very-very unrealistic minimum. That's why I liked your version of 2.5%. It is fair enough for a starting rough estimation I would say.

Reply:
2.5 % comparing to 0.25 % can make a huge different on big RAMs. I have a machine with 512GB so I'll test 0.25% as well. Right now I see that memory.max_usage_in_bytes is smaller than memory.limit_in_bytes by a pretty big number. In reality delta should be small. "limit" is what we set and "usage" must be a bit smaller than limit since we do not have neither overcommit nor ballooning we preallocate whole RAM chunk and VM cannot actually get more memory while running except for caches/buffers

Reply:
I don't think the `memory.usage_in_bytes` value somehow relates to what we are discussing in the scope of the PR. Nevertheless, it's interesting.
The delta between usage and limit depends on how we get the usage... I can imagine the cases when usage is really smaller than the limit, for example, if we define it by counting how much memory is mapped to the guest kernel and userspace...

Reply:
In any case, I would not leave 0.25% as a heuristic here. It's a theoretical minimum. Let it be twice more at least, 0.5% for example...